moodle/Sniffs/Commenting/MissingDocblockSniff.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+// This file is part of Moodle - https://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANdTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <https://www.gnu.org/licenses/>.
+
+namespace MoodleHQ\MoodleCS\moodle\Sniffs\Commenting;
+
+use MoodleHQ\MoodleCS\moodle\Util\Docblocks;
+use MoodleHQ\MoodleCS\moodle\Util\Tokens;
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+
+/**
+ * Checks that all files an classes have appropriate docs.
+ *
+ * @copyright  2024 Andrew Lyons <andrew@nicols.co.uk>
+ * @license    https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class MissingDocblockSniff implements Sniff
+{
+    /**
+     * Register for open tag (only process once per file).
+     */
+    public function register() {
+        return [
+            T_OPEN_TAG,
+        ];
+    }
+
+    /**
+     * Processes php files and perform various checks with file.
+     *
+     * @param File $phpcsFile The file being scanned.
+     * @param int $stackPtr The position in the stack.
+     */
+    public function process(File $phpcsFile, $stackPtr) {
+        $tokens = $phpcsFile->getTokens();
+
+        // Each class, interface, trait, and enum must have a docblock.
+        // If a file has one class, interface, trait, or enum, the file docblock is optinoal.
+        // Otherwise, the file docblock is required.
+
+        // First find out how many items there are.
+        $find = [
+            // Classes, interfaces, and traits.
+            T_CLASS,
+            T_INTERFACE,
+            T_TRAIT,
+
+            // Functions outside of those.
+            T_FUNCTION,
+        ];
+
+        if (version_compare(PHP_VERSION, '8.1.0') >= 0) {
+            $find[] = T_ENUM;
+        }
+
+        $artefactCount = 0;
+        $missingDocblocks = [];
+
+        $typePtr = $stackPtr + 1;
+        while ($typePtr = $phpcsFile->findNext($find, $typePtr + 1)) {
+            $token = $tokens[$typePtr];
+            if ($token['code'] === T_FUNCTION && !empty($token['conditions'])) {
+                // Skip methods of classes, traits and interfaces.
+                continue;
+            }
+
+            $artefactCount++;
+
+            if ($docblock = Docblocks::getDocBlock($phpcsFile, $typePtr)) {
+                // There should be no empty lines between the artefact and the docblock.
```

Comment:
Side comment: For some reason we have been using `artifact` always. I see (didn't know) that `artefact` is the British work. Maybe we should move all cases to the later (not in this PR, just saying).

Reply:
Hah - taht explains a few things. I thought the i version were just typos. Happy to stay consistent for now and go to artifact.